local-dev-environment/search_client.py

```diff
@@ -0,0 +1,117 @@
+import logging
+from datetime import datetime
+from pathlib import Path
+
+import xlsxwriter
+from opensearchpy import OpenSearch
+
+from ingestion_pipeline.config import settings
+from ingestion_pipeline.embedding.embedding_generator import EmbeddingGenerator
+
+# --- 1. CONFIGURE YOUR RUNNING LOCALSTACK OPENSEARCH CONNECTION ---
+# These values should match your LocalStack setup
+
+HOST = "localhost"
+PORT = 9200
+USER = "admin"
+PASSWORD = "really-secure-passwordAa!1"
+CHUNK_INDEX_NAME = settings.OPENSEARCH_CHUNK_INDEX_NAME
+
+# --- 2. Choose your search term and variables for testing ---
+
+SEARCH_TERM = "medicine"
+K_QUERIES = 200  # Number of nearest neighbors to retrieve
+SCORE_FILTER = 0.4  # Minimum score threshold for filtering results
+
+#  Set either boost as zero to return only the other type of search
+KEYWORD_BOOST = 1  # Boost factor for keyword matching in hybrid search
+SEMANTIC_BOOST = 2  # Boost factor for semantic vector search in hybrid search
+
+# --- 3. DEFINE K-NN SEARCH QUERY ---
+
+
+def create_hybrid_query(query_text, query_vector, k=5):
+    return {
+        "size": k,
+        "_source": ["document_id", "page_number", "chunk_text", "case_ref"],
+        "query": {
+            "bool": {
+                "should": [
+                    {"match": {"chunk_text": {"query": query_text, "boost": KEYWORD_BOOST}}},
+                    {"knn": {"embedding": {"vector": query_vector, "k": k, "boost": SEMANTIC_BOOST}}},
+                ]
+            }
+        },
+    }
+
+
+# --- 4. Execute results and write to Excel ---
+def local_search_client():
+    logging.basicConfig(level=logging.INFO)
+    logger = logging.getLogger("embedding_search_client")
+
+    embedding_generator = EmbeddingGenerator(settings.BEDROCK_EMBEDDING_MODEL_ID)
+    embedding = embedding_generator.generate_embedding(SEARCH_TERM)
+    logger.info(f"Generated embedding for search term: '{SEARCH_TERM}'")
+
+    client = OpenSearch(
```

Comment:
Error Handling

There is no error handling for connection failures, search errors, or file writing issues. Consider wrapping OpenSearch and file operations in try/except blocks and logging errors.

Reply:
Create a custom exception for Opensearch and add a helpful error message whilst preserving the original stack trace.

Message example: "The Opensearch Vector Database is down, is it running?"